src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,895 @@
+        }
+
+        Ok(())
+    }
+
+    async fn update_state_if_changed_forward_stage(
+        &mut self,
+        target_splits: Vec<SplitImpl>,
+        current_splits: &mut HashSet<SplitId>,
+        should_trim_state: bool,
+    ) -> StreamExecutorResult<()> {
+        let target_splits: HashSet<SplitId> = target_splits
+            .into_iter()
+            .map(|split| (split.id()))
+            .collect();
+
+        let mut split_changed = false;
+
+        // Checks added splits
+        for split_id in &target_splits {
+            if !current_splits.contains(split_id) {
+                split_changed = true;
+
+                let backfill_state = self
+                    .backfill_state_store
+                    .try_recover_from_state_store(split_id)
+                    .await?;
+                match backfill_state {
+                    None => {
+                        // Newly added split. We don't need to backfill!
+                    }
+                    Some(backfill_state) => {
+                        // Migrated split. It should also be finished since we are in forwarding stage.
+                        match backfill_state {
+                            BackfillState::Finished => {}
+                            _ => {
+                                return Err(anyhow::anyhow!(
+                                    "Unexpected backfill state: {:?}",
+                                    backfill_state
+                                )
+                                .into());
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Checks dropped splits
+        for existing_split_id in current_splits.iter() {
+            if !target_splits.contains(existing_split_id) {
+                tracing::info!("split dropping detected: {}", existing_split_id);
+                split_changed = true;
+            }
+        }
+
+        if split_changed {
+            tracing::info!(
+                actor_id = self.actor_ctx.id,
+                target_splits = ?target_splits,
+                "apply split change"
+            );
+
+            let dropped_splits =
+                current_splits.extract_if(|split_id| target_splits.get(split_id).is_none());
+
+            if should_trim_state {
+                // trim dropped splits' state
+                self.backfill_state_store.trim_state(dropped_splits).await?;
+            }
+
+            self.backfill_state_store
+                .set_states(
+                    target_splits
+                        .into_iter()
+                        .map(|split_id| (split_id, BackfillState::Finished))
+                        .collect(),
+                )
+                .await?;
+        }
+
+        Ok(())
+    }
+}
+
+fn compare_kafka_offset(a: &str, b: &str) -> Ordering {
+    let a = a.parse::<i64>().unwrap();
+    let b = b.parse::<i64>().unwrap();
+    a.cmp(&b)
+}
```

Comment:
Oh. This seems not scalable to connectors other than Kafka. 😄

Reply:
I think this is the only kafka specific logic. We can dispatch